### ISSUE TITLE
Hapi

### ISSUE THREAD

**Issue by brillout** [brillout/vite-plugin-ssr#366]
### Description

From Discord https://discord.com/channels/815937377888632913/815937377888632916/997570716557905941:

> I've been trying (and failing) to integrate vps with a hapi project, and was wondering if anyone had any working examples out there?
>
> My main struggle is the dev server. Prod build is fine. The boilerplate examples (understandably) use Express, and in the case of a dev server it says to register viteDevServer.middlewares as an app-wide middleware. The issue there is that Hapi doesn't use Connect middleware. I'd thought to create a plugin with an onResponse extension point, but hapi's life cycle method arguments of (request, h) don't function at all like (req, res, next).
>
> Can anyone shed some light on whether vps can in fact work with Hapi? Or is it limited to Connect middleware implementations?

**Comment by brillout** [comment]
It's a Vite thing really, vps itself does work with Hapi as you can see in production. Seems like not many people have tried to use Vite's dev middleware with Hapi: https://github.com/vitejs/vite/issues?q=hapi.

I don't know how to make a Connect middleware work for Hapi.

What you can do is to have two Node.js servers in dev: one using Vite's dev server (`$ vite dev`) while the other being your Hapi backend. In prod you only have one Hapi server.

**Comment by BrettLargent** [comment]
@brillout, do you have any suggestions on how to actually go about implementing the vite dev server alongside the node server? I imagine there's some amount of special vite configuration that will still need to be done as just calling vite dev doesn't appear to do anything.

**Comment by BrettLargent** [comment]
The best implementation I've been able to come up with so far is to have two scripts run in parallel - `vite build --ssr --watch` and `nodemon ./server`. It's not as snappy as using the vite middleware but it gets about 95% of the way there. It just requires a manual page refresh or a page navigation to pick up changes, as well as the additional nodemon dev dependency. Ultimately it won't matter once it goes to production so I think this will do. But it'd still be nice to have a more seamless integration someday.

**Comment by brillout** [comment]
In dev you have two processes:
 - The `$ vite` process (Vite's development server) which automatically includes vps (make sure to use the latest vps major version 0.4). Alternatively you can use an Express.js development server, but that shouldn't be required; the Vite development server should be enough.
 - The Hapi server without Vite nor vps (e.g. it does *not* call `renderPage()`)

This means you get a full-fledged Vite frontend including HMR using Vite's built-in development sever, while your Hapi server, in development mode, only serves the backend.

In production:
 - No Vite server at all.
 - Only your Hapi server *without* Vite but *with* vps (e.g. it *does* call `renderPage()`)

In production your Hapi server also serves the frontend: the static files at `dist/client/` and server-side rendering with vps's `renderPage()`.

This setup should work. At least, I don't see any reason why it shouldn't.

**Comment by BrettLargent** [comment]
Ah, I didn't realize that I had an older version of vps installed. Your
solution works as expected!

Thanks for your help!

**Comment by brillout** [comment]
Also see https://github.com/hapijs/hapi/issues/4369.
